### PR TITLE
Bug : Fixes #20911

### LINKED ISSUE #20911: DataFrame.clip() bug when bound a frame when columns not sorted
#### Code Sample, a copy-pastable example if possible

```python
In [14]: import pandas as pd

In [15]: df1 = pd.DataFrame([[1., 0.], [3., 0.]], columns=['A', 'B'])

In [16]: df2 = pd.DataFrame([[100., 1.], [100., 2.]], columns=['B', 'A'])

In [17]: df1.clip(lower=0, upper=df2)
Out[17]:
       A    B
0    1.0  0.0
1  100.0  0.0
In [18]: df1
Out[18]:
     A    B
0  1.0  0.0
1  3.0  0.0

# selecting columns in order produces the correct results
In [20]: df1.clip(lower=0, upper=df2[df1.columns])
Out[20]:
     A    B
0  1.0  0.0
1  2.0  0.0


```
#### Problem description

The behavior was correct in 0.18.1.  df1.A = 3.0 is being clipped to 100.0.  This not expected.

#### Expected Output
```python
In [39]: df1.clip(lower=0, upper=df2)
Out[39]:
     A    B
0  1.0  0.0
1  2.0  0.0
```
#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.14.final.0
python-bits: 64
OS: Linux
OS-release: 3.10.0-327.36.3.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: C
LANG: en_US.UTF-8
LOCALE: None.None

pandas: 0.22.0
pytest: 3.5.0
pip: 9.0.3
setuptools: 39.0.1
Cython: 0.28.2
numpy: 1.14.2
scipy: 1.0.1
pyarrow: 0.9.0
xarray: 0.10.2
IPython: 5.6.0
sphinx: 1.7.2
patsy: 0.5.0
dateutil: 2.7.2
pytz: 2018.4
blosc: None
bottleneck: 1.2.1
tables: 3.4.2
numexpr: 2.6.4
feather: None
matplotlib: 2.2.2
openpyxl: 2.5.2
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.0.2
lxml: 4.2.1
bs4: 4.3.2
html5lib: 0.999
sqlalchemy: 1.2.6
pymysql: None
psycopg2: 2.7.4 (dt dec pq3 ext lo64)
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by cgangwar11**
- [x] closes #20911
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

**Before**
```python
In [1]: import pandas as pd                                                                   

In [2]: df1 = pd.DataFrame([[1., 0.], [3., 0.]], columns=['A', 'B']) 
   ...:                                                                                       

In [3]: df2 = pd.DataFrame([[100., 1.], [100., 2.]], columns=['B', 'A']) 
   ...:                                                                                       

In [4]: df1                                                                                   
Out[4]: 
     A    B
0  1.0  0.0
1  3.0  0.0

In [5]: df2                                                                                   
Out[5]: 
       B    A
0  100.0  1.0
1  100.0  2.0

In [6]: df1.clip(lower=0,upper=df2)                                                           
Out[6]: 
       A    B
0    1.0  0.0
1  100.0  0.0

```
**After**
```python
In [5]: df1.clip(lower=0,upper=df2)                                                           
Out[5]: 
     A    B
0  1.0  0.0
1  2.0  0.0


```

### CODE REVIEW

File: pandas/tests/frame/test_analytics.py [hunk 1]
```diff
@@ -1964,6 +1964,13 @@ def test_clip_against_frame(self, axis):
         tm.assert_frame_equal(clipped_df[lb_mask], lb[lb_mask])
         tm.assert_frame_equal(clipped_df[ub_mask], ub[ub_mask])
         tm.assert_frame_equal(clipped_df[mask], df[mask])
+         # GH 20911, clipping now preserves types
```
Comment (jreback) [inline pandas/tests/frame/test_analytics.py]: make this a new test

File: pandas/tests/frame/test_analytics.py [hunk 2]
```diff
@@ -1964,6 +1964,13 @@ def test_clip_against_frame(self, axis):
         tm.assert_frame_equal(clipped_df[lb_mask], lb[lb_mask])
         tm.assert_frame_equal(clipped_df[ub_mask], ub[ub_mask])
         tm.assert_frame_equal(clipped_df[mask], df[mask])
+         # GH 20911, clipping now preserves types       
+        df1 = DataFrame(np.random.randn(1000,4), columns=['A', 'B','C','D'])
+        df2 = DataFrame(np.random.randn(1000,4), columns=['D', 'A','B','C'])
```
Comment (jreback) [inline pandas/tests/frame/test_analytics.py]: you have some lint errors.

File: pandas/tests/frame/test_analytics.py [hunk 3]
```diff
@@ -1964,6 +1964,13 @@ def test_clip_against_frame(self, axis):
...
+        
+        res1 = df1.clip(lower=0,upper=df2)
```
Comment (jreback) [inline pandas/tests/frame/test_analytics.py]: use result= and expected=

**Review by jreback — CHANGES_REQUESTED** [review]
pls add a whatsnew note

File: pandas/tests/frame/test_analytics.py (hunk 3 above)
Comment (cgangwar11) [reply pandas/tests/frame/test_analytics.py]: will do pardon me for such a rookie mistake....

File: pandas/tests/frame/test_analytics.py [hunk 4]
```diff
@@ -1965,6 +1965,14 @@ def test_clip_against_frame(self, axis):
         tm.assert_frame_equal(clipped_df[ub_mask], ub[ub_mask])
         tm.assert_frame_equal(clipped_df[mask], df[mask])
 
+    def test_clip_against_unordered_columns(self):
+        # GH 20911
+        df1 = DataFrame(np.random.randn(1000, 4), columns=['A', 'B', 'C', 'D'])
+        df2 = DataFrame(np.random.randn(1000, 4), columns=['D', 'A', 'B', 'C'])
+        result = df1.clip(lower=0, upper=df2)
```
Comment (jreback) [inline pandas/tests/frame/test_analytics.py]: can you try this with lower as well

**Review by jreback — CHANGES_REQUESTED** [review]
pls add a whatsnew entry

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm ping on green

**Comment by cgangwar11** [comment]
### ***Better Example***
**Before`**
```python
In [11]: import pandas as pd
    ...: 
    ...: df1 = pd.DataFrame([[50., 60.,70.], [80., 90.,100.]], columns=['A', 'B','C'])
    ...: 
    ...: upper = pd.DataFrame([[50., 80.,49.], [100.,100.,90.]], columns=['B', 'C','A'])
    ...: 
    ...: lower = pd.DataFrame([[60., 40.,35.], [90.,95.,85.]], columns=['C', 'B','A'])
    ...: 
    ...: 

In [12]: df1
Out[12]: 
      A     B      C
0  50.0  60.0   70.0
1  80.0  90.0  100.0

In [16]: lower[df1.columns]
Out[16]: 
      A     B     C
0  35.0  40.0  60.0
1  85.0  95.0  90.0

In [17]: upper[df1.columns]
Out[17]: 
      A      B      C
0  49.0   50.0   80.0
1  90.0  100.0  100.0

In [18]: df1.clip(lower=lower,upper=upper)
Out[18]: 
      A     B      C
0  50.0  80.0   70.0
1  90.0  95.0  100.0

```

**After**
```python

In [8]: (df1.clip(lower=lower, upper=upper))
Out[8]:
      A     B      C
0  49.0  50.0   70.0
1  85.0  95.0  100.0
```

**Comment by cgangwar11** [comment]
@jreback it's green now

**Comment by jreback** [comment]
thanks @cgangwar11

**Pull request merged**